src/lennybot/check/docker_image_available.py

```diff
@@ -66,20 +77,61 @@ def _parse_image(self):
             return DockerImage(None, match.group(2) + "/" + match.group(3), image_tag)
         return DockerImage(match.group(4), match.group(5) + "/" + match.group(6), image_tag)
 
+    def _authenticate_on_registry(self, registry: str, realm: str, service: str, scope: str) -> str:
+        if registry not in self._container_config.registries.keys():
+            raise Exception(f"No credentials found for registry: {registry}")
+
+        registry_data = self._container_config.registries[registry]
+
+        if registry_data.password is None or "":
+            url = f"{realm}?scope={scope}&grant_type=password&service={service}&username={registry_data.username}&password={registry_data.password}&client_id=lennybot&access_type=offline"
+            response = requests.get(url)
+        else:
+            url = f"{realm}?scope={scope}&grant_type=password&service={service}&client_id=lennybot&access_type=offline"
+            response = requests.get(url)
+
+        if response.status_code == 401:
+            raise Exception("Error occured: Unauthenticated: ", response.status_code)
+
+        if response.status_code == 200:
+            token_data = response.json()
+            access_token = token_data.get("token")
+            return access_token
+
+        ## TODO: test for unauthenticated
+
     def _exists_on_docker_hub(self, image: DockerImage):
-        res = requests.get(f"https://hub.docker.com/v2/repositories/{image._name}/tags?page_size=10000")
+        url = f"https://hub.docker.com/v2/repositories/{image._name}/tags?page_size=10000"
+        res = requests.get(url)
         if res.status_code != 200:
-            raise Exception(f"Unexpected status: {res.status_code}")
+            raise Exception(f"Unexpected status: {res.status_code} for url: {url}")
 
         for tag in res.json()["results"]:
             if tag["name"] == image._tag:
                 return True
         return False
 
-    def _exists_on_registry(self, image: DockerImage):
-        res = requests.get(f"https://{image._registry}/v2/{image._name}/manifests/{image._tag}")
+    def _exists_on_registry(self, image: DockerImage, access_token: Optional[str] = None) -> bool:
+        request_url = f"https://{image._registry}/v2/{image._name}/manifests/{image._tag}"
+
+        if access_token is not None:
+            headers = {"Authorization": f"Bearer {access_token}"}
+            res = requests.get(request_url, headers=headers)
+        else:
+            res = requests.get(request_url)
+
+        if res.status_code == 401 and access_token is None:
+            registry = image._registry
+            authenticate = res.headers["Www-Authenticate"]
+            realm = authenticate.split('realm="')[1].split('"')[0]
+            service = authenticate.split('service="')[1].split('"')[0]
+            scope = authenticate.split('scope="')[1].split('"')[0]
```

Comment:
Create Authentication Class and parsing function which returns an object of this class